File: backend/src/main/java/com/mysite/knitly/domain/order/entity/OrderItem.java
package com.mysite.knitly.domain.order.entity;

import com.mysite.knitly.domain.product.product.entity.Product;
import jakarta.persistence.*;
import lombok.*;

@Entity
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@AllArgsConstructor
@Builder
@Table(name = "order_items")
public class OrderItem {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long orderItemId;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "order_id")
    private Order order;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "product_id")
    private Product product;

    @Column(nullable = false)
    private Double orderPrice; // 주문 당시의 상품 가격

    @Column(nullable = false)
    private int quantity; // 수량 (항상 1)

    public OrderItem(Order order, Product product, Double orderPrice, int quantity) {
        this.order = order;
        this.product = product;
        this.orderPrice = orderPrice;
        this.quantity = quantity;
    }

    //== 연관관계 편의 메서드용 Setter ==//
    public void setOrder(Order order) {
        this.order = order;
    }
}

/*
CREATE TABLE `order_items` (
        `order_item_id` BIGINT          NOT NULL    AUTO_INCREMENT,
    `order_id`      BIGINT          NOT NULL,
        `product_id`    BIGINT          NOT NULL,
        `order_price`   DECIMAL(10, 2)  NOT NULL,  -- 주문 당시의 개별 상품 가격
    `quantity`      INT             NOT NULL,  -- 주문 수량 (항상 1)
PRIMARY KEY (`order_item_id`),
FOREIGN KEY (`order_id`) REFERENCES `orders` (`order_id`),
FOREIGN KEY (`product_id`) REFERENCES `products` (`product_id`)
);
*/

File: backend/src/main/java/com/mysite/knitly/domain/order/entity/OrderItem.kt
package com.mysite.knitly.domain.order.entity

import com.mysite.knitly.domain.product.product.entity.Product
import jakarta.persistence.*

@Entity
class OrderItem(
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "product_id", nullable = false)
    val product: Product,

    @Column(nullable = false)
    val orderPrice: Double,

    @Column(nullable = false)
    val quantity: Int
) {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val orderItemId: Long? = null

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "order_id")
    var order: Order? = null
        protected set

    fun assignOrder(order: Order) {
        if (this.order != null) {
            throw IllegalStateException("이미 주문이 할당되었습니다.")
        }
        this.order = order
    }
}